python/ffsim/tenpy/circuits/gates.py

```diff
@@ -0,0 +1,419 @@
+        NNgate = NNgate_a @ NNgate_b
+    else:
+        raise ValueError("undefined spin")
+
+    # convert to (N, Sz)-symmetry-conserved basis
+    NNgate_sym = sym_cons_basis(NNgate)
+
+    return NNgate_sym
+
+
+def apply_gate1(U1: np.ndarray, site: int, psi: MPS) -> None:
+    r"""Apply a single-site gate to a
+    `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+    wavefunction.
+
+    Args:
+        U1: The single-site quantum gate.
+        site: The gate will be applied to `site` on the
+            `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        psi: The `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+
+    Returns:
+        None
+    """
+
+    # on-site
+    U1_npc = npc.Array.from_ndarray(U1, [shfs.leg, shfs.leg.conj()], labels=["p", "p*"])
+    psi.apply_local_op(site, U1_npc)
+
+
+def apply_gate2(
+    U2: np.ndarray,
+    site: int,
+    psi: MPS,
+    eng: TEBDEngine,
+    chi_list: list,
+    norm_tol: float = 1e-5,
+) -> None:
+    r"""Apply a two-site gate to a `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+    wavefunction.
+
+    Args:
+        U2: The two-site quantum gate.
+        site: The gate will be applied to `(site-1, site)` on the `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        psi: The `TeNPy MPS <https://tenpy.readthedocs.io/en/latest/reference/tenpy.networks.mps.MPS.html#tenpy.networks.mps.MPS>`__
+            wavefunction.
+        eng: The
+            `TeNPy TEBDEngine <https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.tebd.TEBDEngine.html#tenpy.algorithms.tebd.TEBDEngine>`__.
+        chi_list: The list to which to append the MPS bond dimensions as the circuit is
+            evaluated.
+        norm_tol: The norm error above which we recanonicalize the wavefunction, as
+            defined in the
+            `TeNPy documentation <https://tenpy.readthedocs.io/en/latest/reference/tenpy.algorithms.dmrg.DMRGEngine.html#cfg-option-DMRGEngine.norm_tol>`__.
+
+    Returns:
+        None
+    """
+
+    # bond between (site-1, site)
+    U2_npc = npc.Array.from_ndarray(
+        U2, [shfsc, shfsc.conj()], labels=["(p0.p1)", "(p0*.p1*)"]
+    )
+    U2_npc_split = U2_npc.split_legs()
+    eng.update_bond(site, U2_npc_split)
+    chi_list.append(psi.chi)
+
+    # recanonicalize psi if below error threshold
+    if np.linalg.norm(psi.norm_test()) > norm_tol:
+        psi.canonical_form_finite()
+
+
+def apply_orbital_rotation(
+    mat: np.ndarray,
+    psi: MPS,
+    eng: TEBDEngine,
+    chi_list: list,
+    norm_tol: float = 1e-5,
+) -> None:
```

Comment:
Here and elsewhere, we should try to match the API of the `ffsim.gates` module as much as possible, e.g. https://github.com/qiskit-community/ffsim/blob/420381f2655fabfa12ed76f7934a968f9536d1e7/python/ffsim/gates/orbital_rotation.py#L44-L51.
Note that the state is always the first argument. In this case, that means `psi` should be the first argument. Also, for now let's make all of the arguments beside `psi` and `mat` keyword-only arguments, in case we want to change something later and not have it be a breaking change.